src/gui/render/graphics_shaders/individual_int_renderer.hpp

```diff
@@ -0,0 +1,406 @@
+#include "gui_render_types.hpp"
+
+#include <GL/glew.h>
+#include <GLFW/glfw3.h>
+#include <glm/glm.hpp>
+#include <glm/gtc/matrix_transform.hpp>
+
+#include <memory>
+#include <vector>
+
+namespace gui {
+
+namespace models {
+
+/**
+ * @brief Renders the meshes to the screen
+ *
+ * @details IndividualIntRenderer renders the meshes given to the screen.
+ * This class handles IndividualInt data structures and classes that derive from
+ * IndividualInt.
+ */
+template <class T>
+class IndividualIntRenderer :
+    public render_to::frame_buffer_multisample,
+    public render_to::frame_buffer,
+    public render_to::shadow_map {
+ protected:
+    GLuint program_id_render_; // ID of render program
+    GLuint program_id_shadow_; // ID of shadow program
+
+    GLuint matrix_id_render_;      // uniform ID of transform matrix
+    GLuint view_matrix_id_render_; // ID of view projection matrix for indexed meshes
+    GLuint depth_bias_id_render_;  // ID of depth projection matrix for indexed meshes
+    GLuint shadow_map_id_render_;  // ID of the shadow map for indexed meshes
+    GLuint color_map_id_render_;   // ID of the color map for indexed meshes
+    GLuint light_direction_id_render_; // ID of the light direction uniform for indexed
+
+    GLuint depth_bias_id_shadow_; // ID of depth projection matrix for indexed meshes
+    // ------ the below are added to the class ------
+    GLuint depth_texture_;              // ID of the shadow depth texture
+    glm::vec3 light_direction_;         // direction of sun light
+    glm::mat4 depth_projection_matrix_; // projection matrix of the light source
+    glm::mat4 depth_view_matrix_; // convert a point in world space to depth in light
+    //  light direction
+    std::vector<std::shared_ptr<T>> meshes_;
+
+ public:
+    /**
+     * @brief Construct a new Main Renderer object
+     *
+     * @param ShaderHandler
+     */
+    IndividualIntRenderer(ShaderHandler shader_handler = ShaderHandler());
+
+    ~IndividualIntRenderer();
+
+    /**
+     * @brief adds a non-indexed mesh so it will cast a shadow
+     *
+     * @param mesh the mesh to add
+     */
+    void add_mesh(std::shared_ptr<T> mesh);
+
+    /**
+     * @brief Set the depth texture ID
+     *
+     * @param texture_id the depth texture ID
+     */
+    void set_depth_texture(GLuint texture_id);
+
+    /**
+     * @brief Set the light direction vector
+     *
+     * @param light_direction the direction of the light
+     */
+    void set_light_direction(glm::vec3 light_direction);
+
+    /**
+     * @brief Set the depth projection matrix
+     *
+     * @param depth_projection_matrix the projection matrix
+     */
+    void set_depth_projection_matrix(glm::mat4 depth_projection_matrix);
+
+    /**
+     * @brief renders the given meshes
+     *
+     * @param window the OpenGL window
+     * @param GLuint frame buffer to render to
+     */
+    int render_frame_buffer(GLFWwindow* window, GLuint frame_buffer = 0) const override;
```

Comment:
what does this return?

Reply:
it was suposed to be a status, but I don't think the function will ever return anything but 0.